### PR TITLE
Allow changing queue names during recovery

### PR DESCRIPTION
**Pull request by vikinghawk**
## Proposed Changes

Allow a consumer to register a callback to change the name of a queue during recovery. This is useful when dealing with transient and auto-delete queues that are susceptible to 404 NOT_FOUND client recovery errors and binding corruption issues during cluster node cycles/upgrades. When a client reconnects and redeclares the queue, we expose a timing issue due to the server trying to cleanup the queue at the same time the client is coming back to redeclare it.

The easy existing alternative is to use server named queues. However, with that consumers have no ability to provide a useful queue name that makes it easy to tell what application a queue belongs too from a monitoring/troubleshooting perspective.

This enhancement allows for applications to still name queues in meaningful ways, but to then support modifying that name during a recovery to avoid the timing issues described above.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [ ] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

### CODE REVIEW

File: src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java [hunk 1]
```diff
@@ -782,11 +790,7 @@ public void recoverQueue(final String oldName, RecordedQueue q, boolean retry) {
                         this.propagateQueueNameChangeToConsumers(oldName, newName);
                         // bug26552:
                         // remove old name after we've updated the bindings and consumers,
-                        // plus only for server-named queues, both to make sure we don't lose
-                        // anything to recover. MK.
-                        if(q.isServerNamed()) {
-                            deleteRecordedQueue(oldName);
-                        }
+                        deleteRecordedQueue(oldName);
```
Comment (vikinghawk) [inline src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java:793]: wanted to call out that this is removing the q.isServerNamed() check here. But i don't think that is an issue as i don't believe their would have historically been anyway for the name to change unless it was server named.

**Review by michaelklishin — APPROVED** [review]

**Comment by michaelklishin** [comment]
Thank you!

**Comment by michaelklishin** [comment]
@vikinghawk can you please submit this PR against `main`?

**Comment by michaelklishin** [comment]
Actually no need to do that, I managed to merge `5.x.x-stable` into `main`.

**Pull request merged**